### ISSUE TITLE
[enhancement] Support ServerOptions inDevice and outDevice in Windows

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#3990]
Environment
-----------

* Your SuperCollider version: 3.9.x
* Your operating system and version: Windows 2010

Expected Behavior
-----------------

Audio driver technologies typically found in Windows (e.g. MME, WASAPI, WDM) tend to publish inputs and outputs as separate devices, for instance:

```
MME : Microsoft Soundmapper - Input (device #0 with 2 ins 0 outs)
MME : Microsoft Soundmapper - Output (device #4 with 0 ins 2 outs)
```

(ASIO is an exception, which tends to show duplex devices.)

The user should be able to select two devices for duplex audio using the same interface supported in MacOS:

```
s.options.inDevice = "MME : Microsoft Soundmapper - Input";
s.options.outDevice = "MME : Microsoft Soundmapper - Output";
```

Probably a couple of restrictions would be needed: both devices must be the same driver type (unless we're certain PortAudio can handle input from MME and output to WASAPI, for instance), and both devices might need to be the same hardware interface. It all depends on what PortAudio can do.

Current Behavior
----------------

`inDevice` and `outDevice` are supported in MacOS only. Effectively, this means that Windows users must either use ASIO (and there are reports that asio4all can be flaky), or otherwise the only way they can choose devices is to set the system default in Windows's control panel.

Caveats: I'm not on Windows myself. If you specify one input or output device using `s.options.device`, *maybe* PortAudio will find the matching device and open them both. That seems pretty unlikely to me. What I do know for sure is that scsynth's `-H` argument can accept only one device name, while we see a *lot* of soundcards split among multiple virtual devices.

Also adding supernova tag (feature parity).

(Note also that it would make no sense to implement this for Linux, because JACK handles the soundcard connection and scsynth has no opportunity to override JACK's hardware device. `-H` is used as the name under which scsynth appears in JACK.)

**Comment by jamshark70** [comment]
> From this i infer that either i've just been stupendously lucky, or the issues under discussion here are particular to Windows 10, or to using some non-standard device.

It's quite possible that the linked issue was because of device contention (some other app may have had exclusive access to the device). Information from the user was not systematic so we may never know.

Regardless, it's a gap in our programming interface that Windows presents separate input and output devices but our software can't select them separately.

**Comment by dyfer** [comment]
@alln4tural I don't think this is a Win10 issue.
It comes up when you either a) need to use an external audio interface and/or b) need low latency operation (from what I've seen the first device chosen by the server is always through the MME driver which usually gives considerable latency).

**Comment by Jomme5** [comment]
Problem on Windows 10 with latest version of Supercolider(3.10)

Got no audio. Can't choose between 'device options' with s.options.inDevice or .outDevice.
Portaudio still make it's own choice.

Solutions?

**Comment by jamshark70** [comment]
Needs more detail.

When you boot the server, it should print a list of devices. Can you copy/paste that list here?

Then, what are the exact `s.options.inDevice` / `...outDevice` commands that you are running?

**Comment by Jomme5** [comment]
Commands:

```
Server.local.options.inDevice = "";
Server.local.options.outDevice = "Windows DirectSound : Luidsprekers (Realtek High Definition Audio)";

Got this:
```

**Comment by Jomme5** [comment]
Saw that these commands only supported by Mac. I'm working on Windows 10. How to change these on Windows 10?

**Comment by dyfer** [comment]
The ability to choose devices properly in Windows is still work-in-progress - choosing independent input/output devices is currently broken.
Your best bet is to use ASIO4ALL IMO - configure it to use the proper devices and then use 
```s.options.device = "ASIO4ALL V2"```

**Comment by Jomme5** [comment]
Works. But Asio4all took exclusively the audio stream. Playing youtube video's go on black and give a error-message. Maybe take a look at lmms.io - multiplatform and no audio issues.

**Comment by snappizz** [comment]
fixed #4475 thanks @dyfer!!!!!

**Comment by brianlheim** [comment]
i think we still need the implementation in supernova before closing this, right?

**Comment by snappizz** [comment]
DAMNIT

**Comment by dyfer** [comment]
I think supernova 1) doesn't prevent separate devices on Windows, 2) doesn't compile on Win with the default toolchain and 3) has other problems which will be addressed in #4476 (mostly it needs to list APIs).

IMO this issue is resolved in #4475, since the language part should allow independent device selection for supernova (although other issues might make it hard regardless). Feel free to reopen if you disagree :)

**Comment by brianlheim** [comment]
doh, completely forgot about that, thanks @dyfer!
